src/pudl/output/eia860.py

```diff
@@ -464,3 +485,616 @@ def ownership_eia860(pudl_engine, start_date=None, end_date=None):
+    many currently "boilerless" CCNG units that use generator ID as boiler ID in
+    the latter years. We could try and apply this more generally, but in cases
+    of generator IDs that haven't been used as boiler IDs, it would break the
+    foreign key relationship with the boiler table, unless we added them there
+    too, which seems like too much deep muddling.
+
+    Args:
+        gens_df (pandas.DataFrame): An generators_eia860 dataframe, which must
+            contain columns: report_date, plant_id_eia, generator_id,
+            unit_id_pudl, bga_source.
+
+    Returns:
+        pandas.DataFrame: with the same columns as the input dataframe, but
+        having some NA values filled in for both the unit_id_pudl and bga_source
+        columns.
+
+    """
+    # forward and backward fill the unit IDs
+    gen_ids = ["plant_id_eia", "generator_id"]
+    gens_df = gens_df.sort_values(["report_date", "plant_id_eia", "generator_id"])
+
+    bfill_units = gens_df.groupby(gen_ids)["unit_id_pudl"].bfill()
+    bfill_idx = (bfill_units.notnull()) & (gens_df.unit_id_pudl.isnull())
+    gens_df.loc[bfill_idx, "bga_source"] = "bfill_units"
+    gens_df.loc[bfill_idx, "unit_id_pudl"] = bfill_units.loc[bfill_idx]
+
+    ffill_units = gens_df.groupby(gen_ids)["unit_id_pudl"].ffill()
+    ffill_idx = (ffill_units.notnull()) & (gens_df.unit_id_pudl.isnull())
+    gens_df.loc[ffill_idx, "bga_source"] = "ffill_units"
+    gens_df.loc[ffill_idx, "unit_id_pudl"] = ffill_units.loc[ffill_idx]
+    gens_df["bga_source"] = gens_df["bga_source"].astype(pd.StringDtype())
+
+    return gens_df
+
+
+def max_unit_id_by_plant(gens_df):
+    """
+    Identify the largest unit ID associated with each plant so we don't overlap.
+
+    The PUDL Unit IDs are sequentially assigned integers. To assign a new ID, we
+    need to know the largest existing Unit ID within a plant. This function
+    calculates that largest existing ID, or uses zero, if no Unit IDs are set
+    within the plant.
+
+    Note that this calculation depends on having all of the pre-existing
+    generators and units still available in the dataframe!
+
+    Args:
+        gens_df (pandas.DataFrame): A generators_eia860 dataframe containing at
+            least the columns plant_id_eia and unit_id_pudl.
+
+    Returns:
+        pandas.DataFrame: Having two columns: plant_id_eia and max_unit_id_pudl
+        in which each row should be unique.
+
+    """
+    return (
+        gens_df[["plant_id_eia", "unit_id_pudl"]]
+        .drop_duplicates()
+        .groupby("plant_id_eia").agg({"unit_id_pudl": max})
+        .fillna(0)
+        .rename(columns={"unit_id_pudl": "max_unit_id_pudl"})
+        .reset_index()
+    )
+
+
+def assign_simple_unit_ids(
+    gens_df,
+    prime_mover_codes,
+    fuel_type_code_pudl=None,
+    label_prefix="single"
+):
+    """
+    Assign unique PUDL Unit IDs to all generators of a given prime mover type.
```

Comment:
this makes it sound like you are lumping generators together

Reply:
Yes you're right. I've re-worded to make it clearer.